Main.hs

```diff
@@ -4,26 +4,87 @@ import Control.Monad (unless)
 import Text.Printf (printf)
 
 data Expr
+  = Num Double
+  | Sqrt Expr
+  | Add Expr Expr
+  | Sub Expr Expr
+  | Mul Expr Expr
+  | Div Expr Expr
+  | Pow Expr Expr
+
 
 instance Show Expr where
-  show = undefined
+  show (Num x) = show x
+  show (Sqrt e) = "sqrt(" ++ show e ++ ")"
+  show (Add e1 e2) = "(" ++ show e1 ++ " + " ++ show e2 ++ ")"
+  show (Sub e1 e2) = "(" ++ show e1 ++ " - " ++ show e2 ++ ")"
+  show (Mul e1 e2) = "(" ++ show e1 ++ " * " ++ show e2 ++ ")"
+  show (Div e1 e2) = "(" ++ show e1 ++ " / " ++ show e2 ++ ")"
+  show (Pow e1 e2) = "(" ++ show e1 ++ " ^ " ++ show e2 ++ ")"
 
 instance Eq Expr where
-  (==) = undefined
+  (Num x)==(Num y)=x==y
+  (Sqrt e1)==(Sqrt e2)=e1==e2
+  (Add e1 e2)==(Add e3 e4)=e1==e3 && e2==e4
+  (Sub e1 e2)==(Sub e3 e4)=e1==e3 && e2==e4
+  (Mul e1 e2)==(Mul e3 e4)=e1==e3 && e2==e4
+  (Div e1 e2)==(Div e3 e4)=e1==e3 && e2==e4
+  (Pow e1 e2)==(Pow e3 e4)=e1==e3 && e2==e4
+  _ == _ = False
```

Comment:
Try to avoid unnecessary parentheses. Also, whenever your implementation of a derivable type class is standard, it's better to derive it.